File: apps/syn-cli/src/syn_cli/commands/org.py
"""Organization management commands — create, list, show, update, delete."""

from __future__ import annotations

from typing import Annotated

import typer
from rich.table import Table

from syn_cli._output import console, format_timestamp, print_error
from syn_cli.commands._api_helpers import api_delete, api_get, api_post, api_put, build_params

app = typer.Typer(
    name="org",
    help="Manage organizations",
    no_args_is_help=True,
)


@app.command("create")
def create(
    name: Annotated[str, typer.Option("--name", "-n", help="Organization name")],
    slug: Annotated[str, typer.Option("--slug", "-s", help="URL-safe slug")],
    created_by: Annotated[str, typer.Option(help="Creator identifier")] = "cli",
) -> None:
    """Create a new organization."""
    data = api_post(
        "/organizations",
        json={"name": name, "slug": slug, "created_by": created_by},
    )
    console.print(f"[green]Organization created:[/green] {data.get('organization_id', '')}")
    console.print(f"  Name: {name}  Slug: {slug}")


@app.command("list")
def list_orgs() -> None:
    """List all organizations."""
    data = api_get("/organizations")

    orgs = data.get("organizations", [])
    if not orgs:
        console.print("[dim]No organizations found.[/dim]")
        return

    table = Table(title=f"Organizations ({data.get('total', len(orgs))})")
    table.add_column("ID", style="cyan", no_wrap=True)
    table.add_column("Name")
    table.add_column("Slug")
    table.add_column("Systems", justify="right")
    table.add_column("Repos", justify="right")
    table.add_column("Created")

    for o in orgs:
        table.add_row(
            o.get("organization_id", "")[:12],
            o.get("name", ""),
            o.get("slug", ""),
            str(o.get("system_count", 0)),
            str(o.get("repo_count", 0)),
            format_timestamp(str(o.get("created_at") or "")),
        )
    console.print(table)


@app.command("show")
def show(
    organization_id: Annotated[str, typer.Argument(help="Organization ID")],
) -> None:
    """Show details of an organization."""
    o = api_get(f"/organizations/{organization_id}")

    console.print(f"[bold]Organization:[/bold] {o.get('name', '')}")
    console.print(f"  ID:       {o.get('organization_id', '')}")
    console.print(f"  Slug:     {o.get('slug', '')}")
    console.print(f"  Systems:  {o.get('system_count', 0)}")
    console.print(f"  Repos:    {o.get('repo_count', 0)}")
    console.print(f"  Created:  {format_timestamp(str(o.get('created_at') or ''))}")


@app.command("update")
def update(
    organization_id: Annotated[str, typer.Argument(help="Organization ID")],
    name: Annotated[str | None, typer.Option("--name", "-n", help="New name")] = None,
    slug: Annotated[str | None, typer.Option("--slug", "-s", help="New slug")] = None,
) -> None:
    """Update an organization's name or slug."""
    if not name and not slug:
        print_error("Provide at least --name or --slug.")
        raise typer.Exit(1)

    body = build_params(name=name, slug=slug)
    api_put(f"/organizations/{organization_id}", json=body)
    console.print(f"[green]Organization updated:[/green] {organization_id}")


@app.command("delete")
def delete(
    organization_id: Annotated[str, typer.Argument(help="Organization ID")],
    force: Annotated[bool, typer.Option("--force", "-f", help="Skip confirmation")] = False,
) -> None:
    """Delete an organization."""
    if not force and not typer.confirm(f"Delete organization {organization_id}?"):
        console.print("[dim]Aborted.[/dim]")
        raise typer.Exit(0)

    api_delete(f"/organizations/{organization_id}")
    console.print(f"[red]Organization deleted:[/red] {organization_id}")

File: apps/syn-cli-node/src/commands/org.ts
/**
 * Organization management commands — CRUD.
 * Port of apps/syn-cli/src/syn_cli/commands/org.py
 */

import { CommandGroup, type CommandDef, type ParsedArgs } from "../framework/command.js";
import { CLIError } from "../framework/errors.js";
import { api, unwrap } from "../client/typed.js";
import { print, printError, printDim, printSuccess } from "../output/console.js";
import { style, BOLD, CYAN, DIM } from "../output/ansi.js";
import { Table } from "../output/table.js";


function reqId(parsed: ParsedArgs): string {
  const id = parsed.positionals[0];
  if (!id) { printError("Missing org-id"); throw new CLIError("Missing argument", 1); }
  return id;
}

const createCommand: CommandDef = {
  name: "create",
  description: "Create a new organization",
  options: {
    name: { type: "string", short: "n", description: "Organization name" },
    slug: { type: "string", short: "s", description: "URL-safe slug" },
  },
  handler: async (parsed: ParsedArgs) => {
    const name = parsed.values["name"] as string | undefined;
    const slug = parsed.values["slug"] as string | undefined;
    if (!name) { printError("Missing --name"); throw new CLIError("Missing option", 1); }

    const body = { name, slug: slug ?? "", created_by: "cli" as const };

    const d = unwrap(await api.POST("/organizations", { body }), "Create organization");
    printSuccess(`Organization created: ${d.organization_id}`);
    print(`  Name: ${d.name ?? name}`);
    if (d.slug) print(`  Slug: ${d.slug}`);
  },
};

const listCommand: CommandDef = {
  name: "list",
  description: "List all organizations",
  handler: async () => {
    const d = unwrap(await api.GET("/organizations"), "List organizations");
    const items = d.organizations ?? [];
    if (items.length === 0) { printDim("No organizations found."); return; }

    const table = new Table({ title: "Organizations" });
    table.addColumn("ID", { style: CYAN });
    table.addColumn("Name");
    table.addColumn("Slug", { style: DIM });
    table.addColumn("Systems", { align: "right" });
    table.addColumn("Repos", { align: "right" });

    for (const o of items) {
      table.addRow(
        o.organization_id,
        o.name,
        o.slug,
        String(o.system_count),
        String(o.repo_count),
      );
    }
    table.print();
  },
};

const showCommand: CommandDef = {
  name: "show",
  description: "Show organization details",
  args: [{ name: "org-id", description: "Organization ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const id = reqId(parsed);
    const d = unwrap(await api.GET("/organizations/{organization_id}", { params: { path: { organization_id: id } } }), "Get organization");
    print(`${style("Organization:", BOLD)} ${d.name}`);
    print(`  ID:      ${d.organization_id}`);
    if (d.slug) print(`  Slug:    ${d.slug}`);
    print(`  Systems: ${d.system_count}`);
    print(`  Repos:   ${d.repo_count}`);
  },
};

const updateCommand: CommandDef = {
  name: "update",
  description: "Update an organization",
  args: [{ name: "org-id", description: "Organization ID", required: true }],
  options: {
    name: { type: "string", short: "n", description: "New name" },
    slug: { type: "string", short: "s", description: "New slug" },
  },
  handler: async (parsed: ParsedArgs) => {
    const id = reqId(parsed);
    const name = parsed.values["name"] as string | undefined;
    const slug = parsed.values["slug"] as string | undefined;
    if (!name && !slug) { printError("Nothing to update. Use --name or --slug."); throw new CLIError("No updates", 1); }

    const body = {
      ...(name ? { name } : {}),
      ...(slug ? { slug } : {}),
    };
    unwrap(await api.PUT("/organizations/{organization_id}", { params: { path: { organization_id: id } }, body }), "Update organization");
    printSuccess(`Organization ${id} updated.`);
  },
};

const deleteCommand: CommandDef = {
  name: "delete",
  description: "Delete an organization",
  args: [{ name: "org-id", description: "Organization ID", required: true }],
  options: {
    force: { type: "boolean", short: "f", description: "Skip confirmation", default: false },
  },
  handler: async (parsed: ParsedArgs) => {
    const id = reqId(parsed);
    if (parsed.values["force"] !== true) {
      printError(`Use --force to confirm deleting organization ${id}`);
      throw new CLIError("Confirmation required", 1);
    }
    unwrap(await api.DELETE("/organizations/{organization_id}", { params: { path: { organization_id: id } } }), "Delete organization");
    printSuccess(`Organization ${id} deleted.`);
  },
};

export const orgGroup = new CommandGroup("org", "Manage organizations");
orgGroup.command(createCommand).command(listCommand).command(showCommand).command(updateCommand).command(deleteCommand);
